src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -124,13 +127,21 @@ public void Reset()
             Debug.Assert(IsCompleted);
             Debug.Assert(_callbacks == null);
             _isCompleted = false;
-            _exceptionInfo = null;
+            _exception = null;
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
         private void ThrowLatchedException()
         {
-            _exceptionInfo.Throw();
+            // Throw a new exception so as not to corrupt stack trace, as may be thrown multiple times
+            if (_exception is OperationCanceledException ex)
+            {
+                // Differentate OperationCanceled exceptions so Cancellation signals are propergated.
+                throw new OperationCanceledException(ex.Message, ex);
+            }
+            else
+            {
+                throw new IOException(_exception.Message, _exception);
```

Comment:
There are no other catch blocks up the stack for specific exception types?  ObjectDisposedException?  SocketException?  etc.?

Reply:
Advice? Need some kind of .Clone approach to maintain the type... e.g.

`_exceptionInfo.CloneAndThrow();`

Reply:
Or maybe
```
_exceptionInfo.Throw(clone: true);
```

Reply:
Depends on who consumes this.  Are callers limited to call sites we control?  Can arbitrary user code be expecting specific types out of this?

The only approaches I'm aware of are:
- Throw the same exception instance, either directly or via ExceptionDispatchInfo
- Create new instances for a known set of types manually, propagating state like Message and wrapping the original as an InnerException
- Testing to see if the object implements ICloneable, but it's very unlikely to
- Using something super heavy-weight, like BinaryFormatter.Serialize/Deserialize, which should work with the majority of exception types

Reply:
:-/

Reply:
Something like that?